src/iminuit/cost.py

```diff
@@ -107,6 +79,141 @@ def _soft_l1_cost(y, ye, ym):
+    ----------
+    n : array-like
+        Observed counts.
+    p : array-like
+        Probabilities for an event to be end up in each bin.
+        Must satisfy sum(p) == 1.
+
+    Returns
+    -------
+    float
+        Cost function value.
+
+    Notes
+    -----
+    The implementation makes the result asymptotically chi2-distributed and
+    keeps the sum small near the minimum, which helps to maximise the numerical
+    accuracy for Minuit.
+
+    The formula is derived from the likelihood ratio with respect to a
+    saturated model (which has mu = n):
+
+    Q = -2 (lnP - lnP')
+    lnP = ln n! - sum_i k_i! + sum_i k_i ln(p_i)
+    lnP' = ln n! - sum_i k_i! + sum_i k_i ln(k_i / n)
+    Q = -2 sum_i k_i (ln(p_i) - ln(k_i / n))
+      = -2 sum_i k_i (ln(mu_i) - ln(n) - ln(k_i) + ln(n))
+      = 2 sum_i k_i (ln(k_i) - ln(mu_i))
+    """
+    n = np.atleast_1d(n)
+    p = np.atleast_1d(p)
+    mu = np.sum(n) * p
+    return 2 * np.sum(n * (_safe_log(n) - _safe_log(mu)))
+
+
+def poisson_chi2(n, mu):
+    """
+    Compute asymptotically chi2-distributed cost for Poisson-distributed data.
+
+    Parameters
+    ----------
+    n : array-like
+        Observed counts.
+    mu : array-like
+        Expected counts.
+
+    Returns
+    -------
+    float
+        Cost function value.
+
+    Notes
+    -----
+    The implementation makes the result asymptotically chi2-distributed and
+    keeps the sum small near the minimum, which helps to maximise the numerical
+    accuracy for Minuit.
+
+    The formula is derived from the likelihood ratio with respect to a
+    saturated model (which has mu = n):
+
+    Q = -2 (lnP - lnP')
+    lnP = sum_i -mu_i + k_i ln(mu_i) - ln(k_i!)
+    lnP' = sum_i -k_i + k_i ln(k_i) - ln(k_i!)
+    Q = 2 sum_i mu_i - k_i + k_i (ln(k_i) - ln(mu_i))
+    """
+    n = np.atleast_1d(n)
+    mu = np.atleast_1d(mu)
+    return 2 * np.sum(mu - n + n * (_safe_log(n) - _safe_log(mu)))
+
+
+def barlow_beeston_lite_chi2(n, mu, mu_var):
+    """
+    Compute asymptotically chi2-distributed cost for template fit.
+
+    Parameters
+    ----------
+    n : array-like
+        Observed counts.
+    mu : array-like
+        Expected counts. This is the sum of the template components scaled
+        with the template fractions.
+    mu_var : array-like
+        Expected variance of expected counts.
+
+    Returns
+    -------
+    float
+        Cost function value.
+
+    Notes
+    -----
+    The implementation makes the result asymptotically chi2-distributed and
+    keeps the sum small near the minimum, which helps to maximise the numerical
+    accuracy for Minuit.
+    """
+    # b_var is trivially 1 for Poisson, but not for weighted templates
+    b_var = mu_var / mu**2
+
+    # need to solve quadratic equation b^2 + (mu var_b - 1) b - n var_b = 0
+    p = mu * b_var - 1
+    q = -n * b_var
+    beta = 0.5 * (-p + np.sqrt(p**2 - 4 * q))
+
+    return poisson_chi2(n, mu * beta) + np.sum((beta - 1) ** 2 / b_var)
```

Comment:
@HDembinski  I think the bug might come from here. 

What I return is the sum of the negative log likelihood, and I do it as follows:
```python
nll = -n * np.log(beta * mu) + beta * mu + 0.5 * (beta - 1.)**2 / b_var
return np.sum(nll)
```
, which is just equation 10 from the [main reference](https://arxiv.org/pdf/1103.0354.pdf).
I don't think it matches up with what you provide here.
As far as I can tell, you return this:
```python
return 2 * np.sum(mu*beta - n + n * (np.log(n) - np.log(mu*beta))) +  np.sum((beta - 1) ** 2 / b_var)
```
The first part of the expression (`poisson chi2`) contains log(n), which is not present the referenced formula,
and in the second part of the expression, you do not divide by 2. Please let me know what you think :)

Reply:
No, that's all ok, you can add constants to the likelihood without changing anything. poisson_chi2 computes 2 * (-log(L)) and therefore I do not divide by 2 in the Gaussian constraint. It is consistent.

Reply:
I see. Makes sense :)